cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -636,13 +676,21 @@ func (s *service) processExits() {
 		// one, assume they've all exited. One of them will be the correct
 		// process.
 		cps := s.running[e.Pid]
+		for _, cp := range cps {
+			if *cp.PendingExecCounter != 0 {
+				s.lifecycleMu.Unlock()
+				// This exit is for an init process, and there are execs we haven't
+				// emitted an exit for. In order to ensure we emit the exec exit
+				// first, skip processing it here and let the `handleStarted` closure
+				// for those execs emit it.
+				continue EXITS
+			}
+		}
 		delete(s.running, e.Pid)
```

Comment:
What about if `len(cps) > 1`? One or more of those processes might have been an exec, or another init process without any outstanding execs.
```suggestion
		var cps, skipped []containerProcess
		for _, cp := range s.running[e.Pid] {
			if *cp.PendingExecCounter != 0 {
				// This exit is for an init process, and there are execs we haven't
				// emitted an exit for. In order to ensure we emit the exec exit
				// first, skip processing it here and let the `handleStarted` closure
				// for those execs emit it.
				skipped = append(skipped, cp)
			} else {
				cps = append(cps, cp)
			}
		}
		if len(skipped) > 0 {
			s.running[e.Pid] = skipped
		} else {
			delete(s.running, e.Pid)
		}
```

Reply:
Thanks! The previous logic was incorrect.